### PR TITLE
Create JSON file and related functionality for replacing hard-coded labels

### LINKED ISSUE #5205: Create JSON file and related functionality to support replacement of hard-coded issue labels
### Overview

- Labels are hard-coded and spread throughout our Github Action files. 
- If a label is changed in the repo, we have to investigate where that label was being used in Github Actions, and update it.
- We need to add a JSON file containing key/value pairs of the labels in the repository, and import it into each GHA file that uses labels

### Action Items

- [x] Create JSON file with key/value pairs for all repository labels (e.g. a "label directory"). This file should include:
  - [x]  Create a `Key` for each label, written in camelCase: e.g. key for "Time Sensitive" would be "timeSensititve", etc.
  - [x]  The `Value` should be an array of the ["label id", "label display name"]
- [x]  Save JSON file to repo in a new subfolder called `github-actions/utils/_data/label-directory.json`
- [x] Write JS file that queries the JSON file and returns the label display name when given the Key  

To be useful and meet the requirements that initiated this issue, the JSON cannot be static. Whenever a label is updated with a new name or when a new label is added to the repo, the change must propagate dynamically and update the JSON file (i.e. JSON needs CRUD.)
- [x] Create a GHA that is triggered `on: label: types: [edited, created, deleted]` with an associated JS file to update and/or add to the "label directory"
- [x] The GHA will need to write **and commit** changes to the new "label-directory"   
- [x] Prepare ER to document new GHA and label lookups in wiki #6454  

Merge Team:
- [x] When merged, remove dependency from #6454 and change status to "New Issue Approval"
### Resources/Instructions

- Summary of HfLA GitHub Actions [here](https://github.com/hackforla/website/wiki/HfLA-GitHub-Actions)
- Workflow files in HfLA repository [here](https://github.com/hackforla/website/tree/gh-pages/.github/workflows)
- For creating the JSON file, see JSON object literals documentation [here](https://www.w3schools.com/js/js_json_objects.asp)
- Issue #4050 and [related DR](https://github.com/hackforla/website/issues/4050#issuecomment-1646282976)
- Issue #6389

### PR DESCRIPTION
**Pull request by t-will-gillis**
Fixes #5205 

### What changes did you make?
  - Created the JSON file with key/value pairs for all repo labels- `label-directory.json` and located it in a new subfolder
  - Populated the JSON file with data extracted from the HfLA repo for all label names + ids (the directory Values) and generated Key camelCased names 
  - Wrote `retrieve-label-directory.js` that returns the actual label display name value when a label key name is provided- note that this is not currently used. It is a necessary file for the next phase which updates the GHA files
  - Wrote GHA `update-label-directory.yml` that is triggered whenever one of the HfLA labels is edited or is added. 
  - Wrote a companion `update-label-directory.js` that determines whether the label is a name-change (such as by adding an emoji) or is a new label, then adds this change to the 'label directory' JSON,
  - In order to commit the changes to the repo, the GHA finishes with the `stefanzweifel/git-auto-commit-action@v5.0.1`
  - Created #6968  
  
Additionally, there are Google Apps Scripts/ Worksheets associated with this:
- See [Copy of Website Labels 'Source of Truth'](https://docs.google.com/spreadsheets/d/1Y1-ba1ITSs6L6Av0mtX0Qy-Oq-tsKP_rwR-AJfuDYCk/edit#gid=0)
### Why did you make the changes (we will use this info to test)?
  - This series of issues was triggered by the study of [adding emoji's](https://github.com/ikatyang/emoji-cheat-sheet) to issue labels. 
  - We determined that we want the freedom to make changes to the labels without the need to track down all instances of each label being used. The intent of the 'label directory' is to be able to refer to a key label such as '2WeeksInactive' and return the actual label name `2 weeks inactive!!!` or `💤 2 weeks inactive` or however we want to change the label, however many times.
  - Also, ultimately other projects with HfLA could be using Website's GHAs, but those project teams might have named their labels differently. The files here are intended to make it easier for other teams to swap out label names in one location (the JSON file).

### Screenshots of Proposed Changes Of The Website  (if any, please do not screen shot code changes)

- Create label: [Log file](https://github.com/t-will-gillis/website/actions/runs/8227846814/job/22496387359) from personal repo showing label creation
  <details><summary>Create label</summary>  

  Screenshot 2024-03-10 215953  

  Screenshot 2024-03-10 220157  

  Screenshot 2024-03-10 220251

  </details>

- Edit label: [Log file](https://github.com/t-will-gillis/website/actions/runs/8227936265/job/22496615677) from personal repo showing label editing
  <details><summary>Edit label</summary>  

  Screenshot 2024-03-10 220737

  Screenshot 2024-03-10 220649

  Screenshot 2024-03-10 220810

  </details>

- Delete label: [Log file](https://github.com/t-will-gillis/website/actions/runs/8227984414/job/22496740032) from personal repo showing label deletion
  <details><summary>Delete label</summary>  

  Screenshot 2024-03-10 221255

  </details>  
  
  ## Notes to PR Reviewers
  - You will need to have a functioning test environment on your local repo. If you have not done this yet, see [Hack for LA's GitHub Actions](https://github.com/hackforla/website/issues/6537#issuecomment-2041147335) to set up, especially Tips 6-8. Tip 8 describes how to copy all of HfLA's labels to your own repo. GHAs are quirky at first so feel free to ask questions.
  - In addition to the 'files changed' in the PR, there are additional changes that you should make to help with testing:

    - In the `update-label-directory.yml` file:

      - Line 14, redirect to your personal repo.
      - Line 18 & 24, the "github-token" needs to be associated with your repo.
    - Note that none of the labels in the `label-directory.json` file will match your repo because the IDs will be a different number. I.e. if you edit your `good first issue`, the automation will see this as a new label- this is OK. 

  - To test this action, you can first create a new label. Go to "Issues" in your repo***, select "New Issue", and choose "Get Started" for any of the templates. Then on the right hand side select the gear icon for "Labels", then select "Edit labels" at the bottom, and finally "New label" to the upper right. Name the issue anything you want, but it might be easier to create a label beginning with an "a", eg. `A new label` or whatever. "Description" and "Color" can be anything you want. Select "Create label" and confirm your label is shown on the "Labels" page.
  - (*** If you don't see "Issues" as a menu option, go to "Settings" and scroll down about halfway. Under "Features", make sure to check the box for "Issues".)
  - When you see that the label is created:
    - Confirm that GitHub Actions was triggered: select "Actions", then check that a workflow named "Update Label Directory" has run. Feel free to explore the log...
    - Confirm that your new label was added to `/_data/label-directory.json`. It should be at the end of the JSON, and it will begin with "NEW" + your title.
  - Repeat the above steps to return to "Edit labels", select your new label, then make some change to the name. Add an [emoji](https://github.com/ikatyang/emoji-cheat-sheet) if you want. Confirm in that "Actions" has run and that the JSON file was updated.
  - Finally, repeat the above steps to delete the label, and confirm in "Actions". The JSON file will still list the label, but the labelId will be '9999999999'. 
  - The tie-in can be seen here for [Google Apps Script /Worksheet](https://docs.google.com/spreadsheets/d/1Y1-ba1ITSs6L6Av0mtX0Qy-Oq-tsKP_rwR-AJfuDYCk/edit#gid=0)
 
 Please feel free to make suggestions...
Wish List and Outstanding Items:

- abstract the url?
- ? Break Google Apps Script functions into modules
- ? Need "Anomaly Report" functional
- ? Need a way to automatically clear deleted Label Objects from JSON *after* review
- ? Have method/ procedure for updating keyNames hopefully without need to edit files separate, or to do a PR  

Completed:  
- **_Done_**  --->   ~Replace link to Apps Script and fix repo~
- **_Done_**  --->   ~Add + " (deleted)" to labelName when deleted~
- **_Done_**  --->    ~Have timeline on the "Worksheet: GitHub Label Updates Log" automatically clear after certain number of days~
- **_Done_**  --->    ~Very important: Since the URL variable in the yml workflow will change every time we need to edit the Google Apps Script: We need the URL to be referenced rather than hard-coded. Should this use a doGet() call?~
- **_Done_**  --->    ~Remove "show context"~
- **_Withdrawn_** ---> ~Should deleted labels be pulled into a separate object?~  

Future (maybe):
- If the label edit does not include the name, the workflow can skip "post to Apps Script"

### CODE REVIEW

**Comment by gaylem** [comment]
Availability: Weekdays 9am - 5pm CT
Review ETA: Wednesday April 14, 5pm CT

**Comment by w1ld-r** [comment]
@marioantonini I connected with Will about this issue and it looks like it's still in draft, so I'm going to remove us as reviewers for now. 

Also leaving this as a note to anyone else who sees this PR and considers picking it up.

**Comment by mSharifHub** [comment]
Hello, I am having issues testing. I added the following 
`
name: Update Label Directory

on:
  label:
    types: [edited, created, deleted]

jobs:
  Update-Label-Directory:
    runs-on: ubuntu-latest
    # if: github.repository == 'hackforla/website'                 << FIX BEFORE MERGING. FIX URL line 40 BELOW
    if: github.repository == 'mSharifHub/website'
    steps:
     - uses: actions/checkout@v4
       with:
        token: ${{ secrets.HFLA_PROJECT_BOARD_TOKEN }}

     - name: Dump context
       env:
         GITHUB_CONTEXT: ${{ secrets.HFLA_PROJECT_BOARD_TOKEN }}
       run: echo "GITHUB_CONTEXT defined"

     - name: Update label directory
       id: update-label-directory
       uses: actions/github-script@v7
       with:
        github-token: ${{ secrets.HFLA_PROJECT_BOARD_TOKEN }}
        script: |
          const script = require('./github-actions/utils/update-label-directory.js')
          const labelPacket = script({g: github, c: context})
          return labelPacket

     # Note the the URL below matches the *current* deployment URL of the Apps Script associated
     # with the 'Source of Truth' Label spreadsheet. (If something is broken, check this link first)
     - name: Send POST request to Google Apps Script
       if: steps.update-label-directory.outputs.result != ''
       run: |
         curl -X POST "https://script.google.com/macros/s/AKfycbyZksNnRWXB6zSJn5QwaPjHPg27NFAzZM6GSpGBb2hkgk09E68LBBZVcdQ2EUVmPjfF/exec" \
           -H "Content-Type: application/json" \
           -d '${{ steps.update-label-directory.outputs.result }}'

     # Before next step, run `git pull` so that branch is current
     - name: Pull latest changes from gh-pages
       run: git pull

     - name: Commit changes to JSON Label Directory
       uses: stefanzweifel/git-auto-commit-action@v5.0.1
       with:
         # Glob pattern of file to commit, and opt. commit message + author
         file_pattern: github-actions/utils/_data/label-directory.json
         commit_message: Update label directory
         commit_author: GitHub Actions Bot <hackforla-bot@hackforla.org>
`

I then created a label but no actions were triggered. 

if you can further explain more how to test it

**Comment by t-will-gillis** [comment]
Hi @mSharifHub Thanks for the comment.

Were you able to copy the new Project Board? I see that you have a message in the Slack channel asking about an error that occurred while you were setting it up. As long as the column names copy over, you should be OK and we can deal with "PR added to project" workflow omission separately. 

I am looking at your `test-5205` repo and have a couple of comments: 
- I don't see that you have "Issues" enabled in your repo. If you haven't yet, go to "Settings" (last menu item at top right of your repo page), then scroll midway down and click "Issues". 
- I can't see but assuming that you created your "HFLA_PROJECT_BOARD_TOKEN"...
- One last thing, if you have not already, you will need to change your default branch to `test-5205` (if this is the correct one)

Once "Issues" is active, the automation should begin to run at least (and might fail still). Let me know if you get to this point and then we can discuss. Thanks!

**Comment by mSharifHub** [comment]
> Hi @mSharifHub Thanks for the comment.
> 
> Were you able to copy the new Project Board? I see that you have a message in the Slack channel asking about an error that occurred while you were setting it up. As long as the column names copy over, you should be OK and we can deal with "PR added to project" workflow omission separately.
> 
> I am looking at your `test-5205` repo and have a couple of comments:
> 
>     * I don't see that you have "Issues" enabled in your repo. If you haven't yet, go to "Settings" (last menu item at top right of your repo page), then scroll midway down and click "Issues".
> 
>     * I can't see but assuming that you created your "HFLA_PROJECT_BOARD_TOKEN"...
> 
>     * One last thing, if you have not already, you will need to change your default branch to `test-5205` (if this is the correct one)
> 
> 
> Once "Issues" is active, the automation should begin to run at least (and might fail still). Let me know if you get to this point and then we can discuss. Thanks!

I was trying to review the PR before commenting. I have configured the token. I have created a branch from the review branch  and created a label under that branch on git but did see actions being called.

File: github-actions/utils/update-label-directory.js [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+  // If label 'edited' but changes do not include 'name', label directory is not updated and workflow exits
+  if (labelAction === 'edited' && !context.payload.changes.name) {
+    console.log(`${breakLine}\n`);
+    console.log(`What was changed:`);
+    console.log(context.payload.changes);
+    console.log(`\n${breakLine}\n`);
+    labelAction = "no update";
+    labelKey = "n/a";
+    message = `Edit to description and/or color only, no updates to JSON or SoT`;
+    console.log(message)
```
[the remark this answers is not in the document]
Comment (t-will-gillis) [reply github-actions/utils/update-label-directory.js]: Added explicit semicolon

**Comment by santiseccovidal** [comment]
Review ETA: 3 PM Thursday
Availability: Monday-Thursday

**Comment by santiseccovidal** [comment]
Updated review schedule!
Review ETA: 3 PM Friday
Availability: Wednesday-Saturday

**Comment by t-will-gillis** [comment]
Hello @santisecco Thank you for your review so far. Please let me know if there is anything that you have questions about, questions, comments, or concerns. Thanks!

**Review by santiseccovidal — DISMISSED** [review]
Hi @t-will-gillis everything looks great.
Tested the automation in my repo and labels are correctly updated, deleted and added.
I liked the idea of setting the labelID to '9999999999' when the label is deleted. I think it's safer and more efficient than deleting it from the JSON file.

Understood that retrieve-label-directory.js is not yet used but will be in the future, also the presence of the wishlist and that the Google sheets doc is updated using Google Apps Script. 
Branches are correct. And your explanation of Why and What changes you made is clear and concise.

Thanks for working on this and answering my questions on the changes throughout the review process.

**Comment by t-will-gillis** [comment]
Final edits are for linking to the live 'Source of Truth' document as well as updating the JSON labels so that the file is current as of today. Bypassing final review since the previous review was full approval.

**Pull request merged**